boards/feather-m0/bat_voltage.c

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright (C)  2024 TU Dresden
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     boards_feather-m0
+ * @ingroup     saul_bat_voltage
+ * @{
+ * @file
+ * @brief       Implementation of battery voltage convert function
+ *
+ * @author      Martine S. Lenders <martine.lenders@tu-dresden.de>
+ *
+ * @}
+ */
+
+#ifdef MODULE_SAUL_BAT_VOLTAGE
+#include "saul/bat_voltage.h"
+
+int16_t saul_bat_voltage_convert(int32_t adc_sample)
+{
+    /* See
+     * https://learn.adafruit.com/adafruit-feather-m0-basic-proto/power-management
+     *
+     * we return in millivolt so we set the reference voltage to 3300
+     * instead of 3.3
+     */
+    return (int16_t)((adc_sample * 2L * 3300L) / 1024L);
```

Comment:
There is a second formula in https://learn.adafruit.com/adafruit-feather-m0-basic-proto/power-management: `(pin.value * 3.3) / 65536 * 2`

Even though, if the resolution is correct, I don't think that should be it, @crasbe can you try with your setup:

```suggestion
    return (int16_t)((adc_sample * 2L * 3300L) / 65536L);
```

But maybe it is best to use your setup to have an output with the real voltage, the ADC readout and then reverse engineer the formular from that.

Reply:
I don't think that this is the right approach. The Feather-M0 has a voltage divider for the battery voltage, so that connecting 3.3V to the BAT input should absolutely not cause the reading to max out.
Also, this would be a reduction of ~64, whereas the real values are lower by a factor of (nearly exactly) 3.3.

Something is wrong with the settings for the ADC, but I didn't feel like diving into that yesterday. I have no experience with the SAMD21 yet.

Reply:
The issue is here:
https://github.com/RIOT-OS/RIOT/blob/584e14bfcee8b6266e9268da6cbdb2ce04c0e854/boards/feather-m0/include/periph_conf.h#L185-L190

The internal 1V reference is selected with a x1 gain. From the meausrements above it is clear that the ADC is out of range at 2V input at the battery pin. With a 1/2 divider, that's 1V => exactly what the reference voltage is.

What *should* be selected is the VCCA/2 reference with a 1/2 gain, so that the input does not exceed the reference voltage.

```diff
/* ADC Default values */ 
#define ADC_PRESCALER                       ADC_CTRLB_PRESCALER_DIV512 

#define ADC_NEG_INPUT                       ADC_INPUTCTRL_MUXNEG_GND 
-#define ADC_GAIN_FACTOR_DEFAULT             ADC_INPUTCTRL_GAIN_1X
-#define ADC_REF_DEFAULT                     ADC_REFCTRL_REFSEL_INT1V
+#define ADC_GAIN_FACTOR_DEFAULT             ADC_INPUTCTRL_GAIN_DIV2
+#define ADC_REF_DEFAULT                     ADC_REFCTRL_REFSEL_INTVCC1
```

<hr/>

The downside of this is that, contrary to the other analog inputs, the battery voltage does not depend on VCCA and therefore will only be as precise as the analog voltage (which is not really precise).

For a precise measurement, we'd have to measure the internal reference with respect to the VCCA.

Essentially this:

V_BAT = 2* (ADC_BAT / 1024) * 3300 * (ADC_Ref / 1024) * (3300 / 1000)

The first part is the same as previously, but then we factor in the ADC Reference voltage correction. The ADC_Ref value comes from 1V * 1/2 measured with a 1.65V reference, so if VCC was at a perfect 3300mV, the expression would be 1.

However I don't think that this mechanism would be supported at the moment and it probably goes too far as well.
Maybe it would be worth adding a note though that the measurement might be unprecise. On the other hand, that's also what all Arduinos do, so... 🤷